### PR TITLE
Feature/ingest folder

### PR DESCRIPTION
**Pull request by Daniel-KM**
Sometime users want to import some hundreds files for one item, for example all images of a scanned manuscript, so to select the directory is a lot simpler.

### CODE REVIEW

**Comment by Daniel-KM** [comment]
The pr is based on previous one because it fixes the same issue, but it can be separated if needed.

File: Module.php [hunk 1]
```diff
@@ -51,4 +63,236 @@ public function handleConfigForm(AbstractController $controller)
...
+     * Working off the "real" base directory and "real" filepath: both must
+     * exist and have sufficient permissions; the filepath must begin with the
+     * base directory path to avoid problems with symlinks; the base directory
+     * must be server-writable to delete the file; and the file must be a
+     * readable regular file or directory.
+     *
+     * @param \SplFileInfo $fileinfo
+     * @return string|null The real file path or null if the file is invalid.
+     *
+     * @todo Factorize with \FileSideload\Media\Ingester\SideloadDir::verifyFileOrDir()
```
Comment (jimsafley) [inline Module.php:266]: Do you plan on finishing this todo?

File: src/Media/Ingester/SideloadDir.php [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+use Omeka\File\TempFileFactory;
+use Omeka\File\Validator;
+use Omeka\Media\Ingester\IngesterInterface;
+use Omeka\Stdlib\ErrorStore;
+use Omeka\Stdlib\Message;
+use Laminas\Form\Element;
+use Laminas\View\Renderer\PhpRenderer;
+
+/**
+ * @todo Factorize with Sideload and avoid to list files and directory in main directory twice.
```
Comment (jimsafley) [inline src/Media/Ingester/SideloadDir.php:16]: Do you plan on finishing this todo?

File: src/Media/Ingester/SideloadDir.php [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        });
+        $iterator = new \RecursiveIteratorIterator($dir);
+        /** @var \SplFileInfo $file */
+        foreach ($iterator as $filepath => $file) {
+            if ($file->isDir()) {
+                if (!$this->hasMoreDirs && $this->verifyFileOrDir($file, true)) {
+                    // There are two filepaths for one dirpath: "." and "..".
+                    $filepath = $file->getRealPath();
+                    // Don't list empty directories.
+                    // TODO Add the count of files for display.
```
Comment (jimsafley) [inline src/Media/Ingester/SideloadDir.php:250]: Do you plan on finishing this todo?

File: src/Media/Ingester/SideloadDir.php [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+     * Working off the "real" base directory and "real" filepath: both must
+     * exist and have sufficient permissions; the filepath must begin with the
+     * base directory path to avoid problems with symlinks; the base directory
+     * must be server-writable to delete the file; and the file must be a
+     * readable regular file or directory.
+     *
+     * @param \SplFileInfo $fileinfo
+     * @return string|null The real file path or null if the file is invalid.
+     *
+     * @todo Factorize with \FileSideload\Module::verifyFileOrDir()
```
Comment (jimsafley) [inline src/Media/Ingester/SideloadDir.php:284]: Do you plan on finishing this todo?

File: src/Media/Ingester/SideloadDir.php [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+        $directory,
+        $deleteFile,
+        $maxFiles,
+        TempFileFactory $tempFileFactory,
+        Validator $validator
+    ) {
+        // Only work on the resolved real directory path.
+        $this->directory = $directory ? realpath($directory) : '';
+        $this->deleteFile = $deleteFile;
+        $this->maxFiles = $maxFiles;
```
Comment (jimsafley) [inline src/Media/Ingester/SideloadDir.php:72]: Please move `$maxFiles` to the end of the signature, like you did with the Sideload ingester.

File: src/Media/Ingester/SideloadDir.php [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+                    $filepath = $file->getRealPath();
+                    // Don't list empty directories.
+                    // TODO Add the count of files for display.
+                    if (!$this->dirHasNoFileAndIsRemovable($filepath)) {
+                        // For security, don't display the full path to the user.
+                        $relativePath = substr($filepath, $lengthDir);
+                        if (!isset($this->listDirs[$relativePath])) {
+                            // Use keys for quicker process on big directories.
+                            $this->listDirs[$relativePath] = null;
+                            if ($this->maxFiles && ++$countDirs >= $this->maxFiles) {
```
Comment (jimsafley) [inline src/Media/Ingester/SideloadDir.php:257]: I'm not sure what's happening but this seems to ignore the $maxFiles configuration. I set my max files to 10 but all files were ingested by this ingester.

File: src/Media/Ingester/SideloadDir.php [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+     * - store_original: (optional, default true) Store the original file?
+     *
+     * {@inheritDoc}
+     */
+    public function ingest(Media $media, Request $request, ErrorStore $errorStore)
+    {
+        $data = $request->getContent();
+
+        // Checks are already done during pre-hydration, but another check is
+        // needed when the ingester is called directly.
```
Comment (jimsafley) [inline src/Media/Ingester/SideloadDir.php:103]: Are these checks necessary? As you mention, they're already done during pre-hydration but when would the ingester be called directly?

File: src/Media/Ingester/SideloadDir.php (hunk 6 above)
Comment (Daniel-KM) [reply src/Media/Ingester/SideloadDir.php:257]: Or another option if needed.

File: Module.php (hunk 1 above)
Comment (Daniel-KM) [reply Module.php]: Removed in last commit.

File: src/Media/Ingester/SideloadDir.php (hunk 2 above)
Comment (Daniel-KM) [reply src/Media/Ingester/SideloadDir.php]: Removed in last commit.

File: src/Media/Ingester/SideloadDir.php (hunk 3 above)
Comment (Daniel-KM) [reply src/Media/Ingester/SideloadDir.php]: Removed in last commit.

File: src/Media/Ingester/SideloadDir.php (hunk 4 above)
Comment (Daniel-KM) [reply src/Media/Ingester/SideloadDir.php]: Removed in last commit.

File: src/Media/Ingester/SideloadDir.php (hunk 5 above)
Comment (Daniel-KM) [reply src/Media/Ingester/SideloadDir.php]: Fixed in last commit.

File: src/Media/Ingester/SideloadDir.php (hunk 6 above)
Comment (Daniel-KM) [reply src/Media/Ingester/SideloadDir.php]: Added in last commit.

**Comment by Daniel-KM** [comment]
I renamed "folder" by "directory" in last commit.

File: Module.php [hunk 8]
```diff
@@ -49,6 +63,237 @@ public function handleConfigForm(AbstractController $controller)
...
+                } else {
+                    $errorStore->addError('ingest_directory', new Message(
+                        'Invalid ingest directory "%s" specified: incorrect path or insufficient permissions', // @translate
+                        $ingestDirectory
+                    ));
+                }
+                continue;
+            }
+
+            $listFiles = $this->listFiles($directory, !empty($dataMedia['ingest_directory_recursively']));
```
Comment (jimsafley) [inline Module.php:167]: Where does `ingest_directory_recursively` come from?

**Comment by jimsafley** [comment]
I notice that the "Directory" select menu does not include the sideload directory itself. Is there a reason for this?

**Comment by jimsafley** [comment]
Let's say I check "Ingest directory recursively" and select a directory that contains no files but contains a subdirectory that contains files, I expect the ingester to sideload the files in the subdirectory. Instead, I'm getting a `Ingest directory "my_directory" is empty.` error.

File: config/module.ini [hunk 9]
```diff
@@ -1,9 +1,9 @@
 [info]
 version = "1.4.0"
 omeka_version_constraint = "^3.0.0"
-name = "File Sideload"
-description = "Add files that are already on your server to items."
-author = "Omeka Team"
+name = "File and Directory Sideload"
```
Comment (jimsafley) [inline config/module.ini:4]: We're not going to change the name of the module, but we're happy to have you attributed as an author.

Comment (Daniel-KM) [reply config/module.ini]: fixed

**Comment by jimsafley** [comment]
Thank you for this important feature!

**Pull request merged**